### ISSUE TITLE
byexamplecookie for output without end-of-line

### ISSUE THREAD

**Issue by saraedum** [byexamples/byexample#158]
The following MarkDown files does not produce the expected output:
`````
```cpp
?: #include <iostream>

?: std::cout << 1;
1
```
`````

I would have expected this to pass, but instead I get:
```
Expected:
1
Got:
1[byexamplecookie]$
```

I understand that to fix this all I need to do is to add a `<< std::endl` but sometimes this is not possible, e.g., when testing a C API that prints something to stdout without adding a newline.

As a workaround, I changed my interpreter to do:
```python
    def _filter_echo(self, options, output_between_prompts):
        lines = byexample.modules.cpp.CPPInterpreter._filter_echo(self, options, output_between_prompts)
        return (l.split('[byexamplecookie]$')[0] for l in lines)
```

Would it make sense to always split at the byexample cookie like this?

**byexample version**
byexample 10.0.0 (Python 3.9.2) from conda-forge.

**Comment by eldipa** [comment]
Thanks @saraedum . Indeed the code assumed that the interpreter prints the prompt (`[byexamplecookie]$`) in its own line.

This is what most of the interpreters do but `cling` does something more weird with the console and `_filter_echo` is a workaround which will not work in all the cases (as you see).

I'm not sure if `split` is the correct solution.

Think in the following (hypothetical) scenario: you have the following line "foo[byexamplecookie]$bar". After the split, you will get "foo". It is correct, it does not have the prompt but it is missing the "bar" string.

We could test for a prompt at the end with `endswith` and remove it if such exists. This solution should work for your case and it will fail for "foo[byexamplecookie]$bar" but at least it will be shown in the output.

Again, this is 100% speculation. I will review the code at night and make some tests and see if there are other corner cases.

Thanks for the bug report!

**Comment by eldipa** [comment]
Okay, I published a new version with the fix (10.0.1). The details of the fix are in 4e0f772 but the short version is that I removed the cookie (`[byexamplecookie]$`) from the end of all the lines.

If we see another spurious cookie in the future, we will know that it is *another* issue in `_filter_echo`.

When you have time to test it, let me know if everything worked well. I will remain open this ticket until then.
